src/fabtrimino.c

```diff
@@ -0,0 +1,250 @@
+{
+    matrix[0][1] = make_square(Z);
+    matrix[1][1] = make_square(Z);
+    matrix[1][2] = make_square(Z);
+    matrix[2][2] = make_square(Z);
+}
+
+void fill_shape_squares(struct square*** matrix, enum shape shape)
+{
+    switch(shape){
+    case T:
+        fill_T_squares(matrix);
+        break;
+    case I:
+        fill_I_squares(matrix);
+        break;
+    case J:
+        fill_J_squares(matrix);
+        break;
+    case L:
+        fill_L_squares(matrix);
+        break;
+    case Z:
+        fill_Z_squares(matrix);
+        break;
+    case S:
+        fill_S_squares(matrix);
+        break;
+    case O:
+        fill_O_squares(matrix);
+        break;
+    default:
+        printf("INVALID SHAPE: %d\n", shape);
+        break;
+    }
+}
+
+struct fabtrimino* fab_make(enum shape shape)
+{
+    struct fabtrimino* fab = calloc(1, sizeof(*fab));
+
+    fab->matrix = calloc(4, sizeof(struct square**));
+    for (int i = 0; i<4; i++)
+    {
+        fab->matrix[i] = calloc(4, sizeof(struct square*));
+    }
+    nullify_squares_matrix(fab->matrix, 4);
+
+    fill_shape_squares(fab->matrix, shape);
+
+    fab->shape = shape;
+    fab->pos = NULL;
+
+    return fab;
+}
+
+void free_square(struct square* square)
+{
+    free(square);
+}
+
+void fab_free(struct fabtrimino* fab)
+{
+    for (int i = 0; i < 4; i++) // TODO: change the 4 to matrix.size once the matrix is a struct
+    {
+        for (int j = 0; j < 4; j++)
+        {
+            if (fab->matrix[i][j] != NULL)
+            {
+                free_square(fab->matrix[i][j]);
+            }
+        }
+        free(fab->matrix[i]);
+    }
+    free(fab->matrix);
+    if (fab->pos != NULL)
+    {
+        free(fab->pos);
+    }
+    free(fab);
+}
+
+struct color {
+    Uint8 r;
+    Uint8 g;
+    Uint8 b;
+};
+
+struct color shape_color(enum shape shape)
+{
+    struct color col;
+    switch(shape) {
+    case T:
+        col.r = 0x80;
+        col.g = 0x00;
+        col.b = 0x80;
+        return col;
+    case I:
+        col.r = 0x00;
+        col.g = 0xFF;
+        col.b = 0xFF;
+        return col;
+    case L:
+        col.r = 0xFF;
+        col.g = 0xA5;
+        col.b = 0x00;
+        return col;
+    case J:
+        col.r = 0x00;
+        col.g = 0x00;
+        col.b = 0xFF;
+        return col;
+    case S:
+        col.r = 0x00;
+        col.g = 0x80;
+        col.b = 0x00;
+        return col;
+    case Z:
+        col.r = 0xFF;
+        col.g = 0x00;
+        col.b = 0x00;
+        return col;
+    case O:
+        col.r = 0xFF;
+        col.g = 0xFF;
+        col.b = 0x00;
+        return col;
+    default:
+        col.r = 0xAA;
+        col.g = 0xBB;
+        col.b = 0xCC;
+        return col;
+    }
+}
+
+void draw_square(struct square* square, struct vector* position, SDL_Renderer* renderer)
+{
+    struct color color = shape_color(square->shape);
+
+    SDL_SetRenderDrawColor(renderer, color.r, color.g, color.b, 0xFF);
```

Comment:
I generally preach paranoia and the checking of all return values in C. I use the following macros to make my life easier:
```

#define sdl_fail(p)                                                     \
    {                                                                   \
        fprintf(stderr, "sdl-fail<%s:%u> %s: %s\n",                     \
                __FILE__, __LINE__, #p, SDL_GetError());                \
        abort();                                                        \
    } while(false)


#define sdl_err(p)                                                      \
    ({                                                                  \
        typeof(p) ret = (p);                                            \
        if (unlikely(ret < 0)) sdl_fail(p);                             \
        ret;                                                            \
    })

#define sdl_ptr(p)                                                      \
    ({                                                                  \
        typeof(p) ret = (p);                                            \
        if (unlikely(!ret)) sdl_fail(p);                                \
        ret;                                                            \
    })
```
used like so:
```
sdl_err(SDL_SetRenderDrawColor(...));
SDL_Window * window = sdl_ptr(SDL_CreateWindow(...));
```